### ISSUE TITLE
raising an exception from datagramReceived should not kill a listening UDP Port

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11188]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)| @radix reported|
|-|-|
|Trac ID|trac#1695|
|Type|enhancement|
|Created|2006-05-02 20:37:33Z|

the fact that readPrecisely is buggy (or whatever) shouldn't mean that the Port should be closed.

```
2006/04/30 19:10 CDT [twisted.names.dns.DNSDatagramProtocol (UDP)] Traceback (most recent call last):
          File "/home/t-names/Twisted/twisted/python/log.py", line 53, in callWithLogger
            return callWithContext({"system": lp}, func, *args, **kw)
          File "/home/t-names/Twisted/twisted/python/log.py", line 38, in callWithContext
            return context.call({ILogContext: newCtx}, func, *args, **kw)
          File "/home/t-names/Twisted/twisted/python/context.py", line 59, in callWithContext
            return self.currentContext().callWithContext(ctx, func, *args, **kw)
          File "/home/t-names/Twisted/twisted/python/context.py", line 37, in callWithContext
            return func(*args,**kw)
        --- <exception caught here> ---
          File "/home/t-names/Twisted/twisted/internet/selectreactor.py", line 139, in _doReadOrWrite
            why = getattr(selectable, method)()
          File "/home/t-names/Twisted/twisted/internet/udp.py", line 126, in doRead
            self.protocol.datagramReceived(data, addr)
          File "/home/t-names/Twisted/twisted/names/dns.py", line 1088, in datagramReceived
            m.fromStr(data)
          File "/home/t-names/Twisted/twisted/names/dns.py", line 1047, in fromStr
            self.decode(strio)
          File "/home/t-names/Twisted/twisted/names/dns.py", line 992, in decode
            header = readPrecisely(strio, self.headerSize)
          File "/home/t-names/Twisted/twisted/names/dns.py", line 153, in readPrecisely
            raise EOFError
        exceptions.EOFError:
 
2006/04/30 19:10 CDT [twisted.names.dns.DNSDatagramProtocol (UDP)] (Port 53 Closed)
2006/04/30 19:10 CDT [twisted.names.dns.DNSDatagramProtocol (UDP)] Stopping protocol <twisted.names.dns.DNSDatagramProtocol instance at 0xb798580c>
```

<details><summary>Searchable metadata</summary>

```
trac-id__1695 1695
type__enhancement enhancement
reporter__radix radix
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1146602253000000 1146602253000000
changetime__1148243604000000 1148243604000000
version__ 
owner__ 
cc__radix,itamarst,glyph,PenguinOfDoom,ghazel,exarkun,teratorn
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

I disagree.

Fix ```readPrecisely```, or fix this specifically in ```twisted.names```; the behavior of uncaught exceptions when they propogate all the way to the reactor is to disable the event source in question.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

Offline discussion lead to a solution which involved changes to the error handling behavior of udp.Port.

Please review in udp-app-error-1695

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to PenguinOfDoom|
|-|-|

Actually, this still fails on iocp.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix set owner to @radix<br>@radix set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix set owner to @exarkun<br>@radix set status to `new`|
|-|-|

Looks good. I notice that you apparently support the exception-handling even for connected cases in IOCP but not in the basic posix UDP support. That's not a big deal, since it's an advertised feature of connected UDP, so go ahead and merge (unless you want to also support it for connected UDP, in which case you should write a test and fix it).


+1

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

Sorry, I misread that IOCP code. It's not the connected-UDP case, it's a deprecated case. I suggest *not* catching the error from the deprecated case.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `closed`|
|-|-|

(In [16891]) Merge udp-app-error-1695

Author: exarkun
Reviewer: radix
Fixes [#11188](https://github.com/twisted/twisted/issues/11188)

This adds try/except handling around calls to datagramReceived.  Exceptions are
handled by logging them.  The port which generated them is no longer disconnected.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
